File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size()-1;
        int breakPoint = 0;
        for(int i=n; i>0; i--) {
            if(nums[i] > nums[i-1]) {
                breakPoint = i;
                break;
            }
        }
        if(breakPoint == 0) {
            sort(nums.begin(), nums.end());
        }
        else {
            int numToSwap = nums[breakPoint-1];
            int min = INT_MAX;
            for(int j=breakPoint; j<=n; j++) {
                if(nums[j]>numToSwap && nums[j]<min) {
                    int temp = nums[j];
                    nums[j] = nums[breakPoint-1];
                    nums[breakPoint-1] = temp;
                }
            }
            sort(nums.begin()+breakPoint, nums.end());
        }
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        // nums = [1, 5, 4, 3, 2]   2 3 1 4 
        int n = nums.length;
        int i=n-2;
        for(; i>=0; i--) {
            if(nums[i]<nums[i+1]) break;
        }
        if(i==-1) {
            reverse(nums, 0, n-1);
        }

        if(i>=0) {
            int j=n-1;
            while(nums[j]<=nums[i]) {
                j--;
            }
            swap(nums, i, j);
            reverse(nums, i+1, n-1);
        }
    }
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    private void reverse(int[] nums, int i, int j) {
        int left=i, right=j;
        while(left<=right) {
            int temp = nums[left];
            nums[left++] = nums[right];
            nums[right--] = temp;
        }
    }
}